portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -41,9 +67,29 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String content = request.getParameter("comment");
     String name = request.getParameter("name");
-    if (content.length() != 0) {
-      Comment comment = name.length() == 0 ? new Comment(content) : new Comment(name, content);
-      commentsList.add(comment);
+    String stringifiedKey = request.getParameter("comment-key");
+
+    if (content != null && name != null && content.length() != 0) {
+      name = name.length() == 0 ? "anonymous" : name;
+      Date timestamp = new Date();
+      Entity commentEntity = new Entity("Comment");
+      commentEntity.setProperty("name", name);
+      commentEntity.setProperty("content", content);
+      commentEntity.setProperty("numLikes", 0);
+      commentEntity.setProperty("timestamp", timestamp);
+      datastore.put(commentEntity);
+    } else if (stringifiedKey != null) {
+      // Patch for numLikes (executes when comment is liked).
+      Key key = KeyFactory.stringToKey(stringifiedKey);
+      try {
+        Entity retrievedComment = datastore.get(key);
+        long numLikes = (long) retrievedComment.getProperty("numLikes");
+        numLikes++;
+        retrievedComment.setProperty("numLikes", numLikes);
+        datastore.put(retrievedComment);
+      } catch (EntityNotFoundException e) {
+        System.out.println("Exception: " + e.getMessage());
```

Comment:
Lets add some nicer logging for these errors -- check out this util class: https://docs.oracle.com/javase/8/docs/api/java/util/logging/Logger.html 
Essentially you have a logger instance per class type: 
private static final Logger logger = Logger.getLogger(DataServlet.class.getName()); 

so then there are different "levels" of logging  - https://docs.oracle.com/javase/8/docs/api/java/util/logging/Level.html because sometimes we just need informational logging vs when exceptions happen so maybe Id put this as warning level 
logger.log(Level.WARNING, "Entity was not found in the datastore: " + e.getMessage())

We want to use standard frameworks for these types of things as opposed to printlns because on a larger scale when lots of different kinds of things are logged (for info or warning or error purposes - standardisation enables users to filter the logs when they need to look at what happened easily:))